bin/proving-service/src/proxy/worker.rs

```diff
@@ -1,75 +1,216 @@
-use std::time::Duration;
+use std::time::{Duration, Instant};
 
 use pingora::lb::Backend;
+use serde::Serialize;
 use tonic::transport::Channel;
-use tonic_health::pb::{
-    HealthCheckRequest, health_check_response::ServingStatus, health_client::HealthClient,
-};
 use tracing::error;
 
-use super::health_check::create_health_check_client;
-use crate::error::ProvingServiceError;
+use super::{metrics::WORKER_UNHEALTHY, status::create_status_client};
+use crate::{
+    commands::worker::ProverType,
+    error::ProvingServiceError,
+    generated::status::{StatusRequest, status_api_client::StatusApiClient},
+};
+
+/// The maximum exponent for the backoff.
+///
+/// The maximum backoff is 2^[MAX_BACKOFF_EXPONENT] seconds.
+const MAX_BACKOFF_EXPONENT: usize = 9;
 
 // WORKER
 // ================================================================================================
 
 /// A worker used for processing of requests.
 ///
 /// A worker consists of a backend service (defined by worker address), a flag indicating wheter
-/// the worker is currently available to process new requests, and a gRPC health check client.
+/// the worker is currently available to process new requests, and a gRPC status client.
 #[derive(Debug, Clone)]
 pub struct Worker {
     backend: Backend,
-    health_check_client: HealthClient<Channel>,
+    status_client: StatusApiClient<Channel>,
     is_available: bool,
+    health_status: WorkerHealthStatus,
+    version: String,
+}
+
+/// The health status of a worker.
+///
+/// A worker can be either healthy or unhealthy.
+/// If the worker is unhealthy, it will have a number of failed attempts.
+/// The number of failed attempts is incremented each time the worker is unhealthy.
+#[derive(Debug, Clone, PartialEq, Serialize)]
+pub enum WorkerHealthStatus {
+    Healthy,
+    Unhealthy {
+        failed_attempts: usize,
+        #[serde(skip_serializing)]
+        first_fail_timestamp: Instant,
+        reason: String,
+    },
 }
 
 impl Worker {
-    /// Creates a new worker and a gRPC health check client for the given worker address.
+    /// Creates a new worker and a gRPC status client for the given worker address.
     ///
     /// # Errors
     /// - Returns [ProvingServiceError::InvalidURI] if the worker address is invalid.
     /// - Returns [ProvingServiceError::ConnectionFailed] if the connection to the worker fails.
     pub async fn new(
-        worker: Backend,
+        worker_addr: String,
         connection_timeout: Duration,
         total_timeout: Duration,
     ) -> Result<Self, ProvingServiceError> {
-        let health_check_client =
-            create_health_check_client(worker.addr.to_string(), connection_timeout, total_timeout)
-                .await?;
+        let backend =
+            Backend::new(&worker_addr).map_err(ProvingServiceError::BackendCreationFailed)?;
+
+        let status_client =
+            create_status_client(worker_addr, connection_timeout, total_timeout).await?;
```

Comment:
nit: `create_status_client()` seems to be used only in this module - would it make sense to move it here (as a helper function)?